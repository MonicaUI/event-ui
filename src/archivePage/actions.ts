import v1PageMigrations from 'nucleus-widgets/migrations/website/v1/pages';

export const LOAD_EVENT_ARCHIVE_PAGE_DATA = 'event-guestside-site/LOAD_EVENT_ARCHIVE_PAGE_DATA';

export const transformAndLoadEventArchivedPageData = (eventArchivePageData: $TSFixMe) => {
  return (dispatch: $TSFixMe, getState: $TSFixMe): $TSFixMe => {
    const { website: existingWebsite } = getState();
    dispatch({
      type: LOAD_EVENT_ARCHIVE_PAGE_DATA,
      payload: {
        website: {
          ...existingWebsite,
          pages: v1PageMigrations(eventArchivePageData.pages),
          layoutItems: eventArchivePageData.layoutItems,
          pluginData: eventArchivePageData.plugins
        }
      }
    });
  };
};
