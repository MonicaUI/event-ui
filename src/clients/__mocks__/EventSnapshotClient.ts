/**
 * This is a mock for EventSnapshotClient.js. The getEventSnapshot function is
 * instrumented with FutureExcutor, which can simulate a delay, such as network
 * latency, when set.
 * To enable the mock, one should add the following 2 lines:
 *  1. const eventSnapshotClient = require('<path-to-client>/EventSnapshotClient').default.prototype;
 *  2. import { FutureExecutor } from '<path-to-client>/__mocks__/EventSnapshotClient';
 * To enable the delay, the user of the mock should:
 *  1. Create the FutureExecutor, e.g. const futureExecutor = new FutureExecutor
 *  2. Invoke eventSnapshotClient.setFutureExecutor(futureExecutor);
 *  3. Invoke eventSnapshotClient.getEventSnapshot, either directly or indirectly
 *  4. Invoke futureExecutor.resolve(ARRAY_INDICATED_ORDER_OF_RESOLUTION, callbackFunction)
 */
import EventSnapshot from '../../../fixtures/EventSnapshot.json';

const eventSnapshot = { ...EventSnapshot.eventSnapshot, version: 'fIQoY_yI8DDWAUEtQw69TRSXXqKFSabH' };
const eventSnapshot2 = { ...EventSnapshot.eventSnapshot, version: 'yIlzAvkFA.DJJDLTvCAu3gkaiLNF29zu' };

const eventSnapshotMap = new Map();

const getLatestEventSnapshot = () => eventSnapshot2;

const getEventSnapshot = (eventId, options = {}) => {
  // @ts-expect-error ts-migrate(2339) FIXME: Property 'version' does not exist on type '{}'.
  const { version } = options;
  let foundEventSnapshot;
  if (version) {
    foundEventSnapshot = eventSnapshotMap.get(version);
  } else {
    foundEventSnapshot = getLatestEventSnapshot();
  }
  return foundEventSnapshot;
};

export class FutureExecutor {
  promises: $TSFixMe;
  resolvers: $TSFixMe;
  constructor() {
    this.promises = [];
    this.resolvers = [];
  }

  execute =
    (method: $TSFixMe): $TSFixMe =>
    (...args: $TSFixMe[]): $TSFixMe => {
      const p = new Promise(resolve => this.resolvers.push(resolve.bind(null, method(...args))));
      this.promises.push(p);
      return p;
    };

  async resolve(taskIds: $TSFixMe, resolveCallback: $TSFixMe): Promise<$TSFixMe> {
    taskIds.forEach(taskId => this.resolvers[taskId]());
    await Promise.all(this.promises);
    this.resolvers = [];
    this.promises = [];
    resolveCallback();
  }
}

export default class EventSnapshotClient {
  futureExecutor: $TSFixMe;
  constructor() {
    this.initialize();
  }

  initialize(): $TSFixMe {
    eventSnapshotMap.clear();
    eventSnapshotMap.set(eventSnapshot.version, eventSnapshot);
    eventSnapshotMap.set(eventSnapshot2.version, eventSnapshot2);
  }

  setFutureExecutor(futureExecutor: $TSFixMe): $TSFixMe {
    this.futureExecutor = futureExecutor;
  }

  getFutureExecutor(): $TSFixMe {
    return this.futureExecutor;
  }

  getLatestEventSnapshot(): $TSFixMe {
    return getLatestEventSnapshot();
  }

  getEventSnapshot(eventId: $TSFixMe, options = {}): $TSFixMe {
    const futureExecutor = this.getFutureExecutor();
    if (futureExecutor) {
      this.setFutureExecutor(null);
      return futureExecutor.execute(getEventSnapshot)(eventId, options);
    }
    return getEventSnapshot(eventId, options);
  }

  isLatestSnapshotVersion(
    eventId: $TSFixMe,
    accountSnapshotVersion: $TSFixMe,
    eventSnapshotVersion: $TSFixMe
  ): $TSFixMe {
    return eventSnapshot2.version === eventSnapshotVersion; // need to add account check
  }
}
