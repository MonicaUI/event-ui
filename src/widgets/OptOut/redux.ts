/* eslint-disable import/no-cycle */

import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Opt-out and opt-in Redux module:
 * Event invitees can click "Opt Out" link in their emails to opt them out all communications from the event account
 */
const GET_OPT_OUT_STATUS = 'event-guestside-site/optOut/GET_OPT_OUT_STATUS';
const OPT_OUT_CONTACT = 'event-guestside-site/optOut/OPT_OUT_CONTACT';

const createGetOptedOutStatusAction = contactInfo => {
  return {
    type: GET_OPT_OUT_STATUS,
    payload: contactInfo
  };
};

export const getOptOutStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    const optedOutContact = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, guestId, contactId);
    dispatch(createGetOptedOutStatusAction(optedOutContact));
  };

export const createOptedOutAction = (contactInfo: $TSFixMe): $TSFixMe => {
  return {
    type: OPT_OUT_CONTACT,
    payload: contactInfo
  };
};

export const optOutAction =
  (optOutStatus: $TSFixMe, fromConfirmPage: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventGuestClient },
      accessToken,
      event: { id: eventId }
    } = getState();
    let {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    if (contactId) {
      inviteeId = null;
    }
    // eslint-disable-next-line import/no-cycle
    const page = (await import(/* webpackChunkName optOut */ './content')).optInOutPageNew;
    if ((fromConfirmPage && optOutStatus) || !fromConfirmPage) {
      const optedOutContact = await eventGuestClient.optOut(
        accessToken,
        eventId,
        inviteeId,
        guestId,
        contactId,
        optOutStatus
      );
      dispatch(createOptedOutAction(optedOutContact));
    }
    dispatch(injectPageSnapshot(page));
  };

export const prepareForOptOutPageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const guestId = queryParams.g;
    const contactId = queryParams.c;
    let page;
    if (!inviteeId && !guestId && !contactId) {
      page = (await import(/* webpackChunkName optOut */ './invalidInviteeContent')).invalidInviteeOptOut;
    } else {
      await dispatch(getOptOutStatus());
      if (getState() && getState().optOut && getState().optOut.optOutStatus) {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).optInOutPage;
      } else {
        // eslint-disable-next-line import/no-cycle
        page = (await import(/* webpackChunkName optOut */ './content')).confirmationPage;
      }
    }
    dispatch(injectPageSnapshot(page));
  };

export const optOutReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_OPT_OUT_STATUS:
    case OPT_OUT_CONTACT:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
