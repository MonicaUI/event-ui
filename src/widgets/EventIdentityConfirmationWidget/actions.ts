import {
  getEventRegistration,
  getVisibleRegistrationTypesByPlannerSelection
} from '../../redux/selectors/currentRegistrant';
import { getRegistrationPageFields } from '../../redux/selectors/currentRegistrationPath';
import { NOT_VISIBLE, READ_ONLY } from 'cvent-question-widgets/lib/DisplayType';
import { isWidgetReviewed } from '../../redux/website/pageContentsWithGraphQL';
import StandardContactFields from 'event-widgets/lib/StandardContactFields/StandardContactFields';
import { defaultRegistrationTypeId } from 'event-widgets/utils/registrationType';
import { CLEAR_REG_CART_INFERRED_FIELDS } from '../../redux/registrationForm/regCart/actionTypes';
import { setIn } from 'icepick';
import { getRegCart } from '../../redux/selectors/shared';
import { saveRegistration } from '../../redux/registrationForm/regCart';
import { runRegistrationTypeChangeValidationsForPrimaryAndGuest } from '../../redux/registrationForm/regCart';
import {
  handleRegistrationTypeSelectionConflict,
  validateUserRegistrationTypeSelection
} from '../../dialogs/selectionConflictDialogs';
import { withLoading } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { getUpdateErrors } from '../../redux/registrationForm/errors';
import { openNoAdmissionItemAvailableForRegistrationTypeDialog, openPrivateEventErrorDialog } from '../../dialogs';
import { merge } from 'lodash';
import { populateVisibleProducts } from '../../redux/visibleProducts';
import { clearOutOfSyncUserSessionFieldsAndUrlQuery } from '../../redux/registrantLogin/actions';
import { ApolloClient } from '@apollo/client';
import type { RootState, AsyncAppThunk } from '../../redux/reducer';

export const isRegTypeReviewed = async (
  state: RootState,
  registrationTypeId: string,
  registrationPathId: string,
  apolloClient: ApolloClient<unknown>
): Promise<boolean> => {
  const widgetTypeName = 'RegistrationType';
  const visibleRegTypes = getVisibleRegistrationTypesByPlannerSelection(state, registrationPathId) || [];
  const isRegTypeWidgetReviewed = await isWidgetReviewed(state, { widgetType: widgetTypeName }, apolloClient);
  return visibleRegTypes.includes(registrationTypeId) && isRegTypeWidgetReviewed;
};

export function clearFields(): AsyncAppThunk {
  return async (
    dispatch,
    getState,
    {
      apolloClient
    }: {
      apolloClient?: ApolloClient<unknown>;
    }
  ) => {
    try {
      const state = getState();
      const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
      const isRegTypeWidgetReviewed = await isRegTypeReviewed(
        state,
        eventRegistration.registrationTypeId,
        eventRegistration.registrationPathId,
        apolloClient
      );
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        const regTypeChangeValidations = await runRegistrationTypeChangeValidationsForPrimaryAndGuest(
          validateUserRegistrationTypeSelection,
          defaultRegistrationTypeId,
          getState(),
          true,
          apolloClient
        );
        // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
        if (regTypeChangeValidations && regTypeChangeValidations.hasConflicts) {
          return dispatch(
            handleRegistrationTypeSelectionConflict(
              regTypeChangeValidations.validationResults,
              null,
              null,
              dissociateInviteeAndClearInferredFields
            )
          );
        }
      }
      const linkedInviteeIdToRemove = eventRegistration.attendee.attendeeId;
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        eventRegistration.registrationTypeId = defaultRegistrationTypeId;
      }
      /*
       * This is an existing pattern that we will need to revisit later. We should clear
       * the individual field rather than replace the entire reg cart
       */
      dispatch({
        type: CLEAR_REG_CART_INFERRED_FIELDS,
        payload: {
          regCart: setIn(
            getRegCart(state),
            ['eventRegistrations', eventRegistration.eventRegistrationId],
            eventRegistration
          )
        }
      });
      /**
       * on clear fields regType is cleared, need to reevaluate visible products
       */
      await dispatch(populateVisibleProducts(eventRegistration.eventRegistrationId));
      await dispatch(dissociateInviteeAndClearInferredFields());
      return await dispatch(withLoading(saveRegistration)(linkedInviteeIdToRemove));
    } catch (error) {
      if (
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 1 arguments, but got 2.
        getUpdateErrors.isPrivateEvent(error, getState()) ||
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 0-1 arguments, but got 2.
        getUpdateErrors.isAttendeeNotAllowedByCustomLogic(error, getState())
      ) {
        return await dispatch(openPrivateEventErrorDialog());
      }
      if (getUpdateErrors.isAdmissionItemsNotAvailableForRegTypeError(error)) {
        return await dispatch(openNoAdmissionItemAvailableForRegistrationTypeDialog());
      }
      throw error;
    }
  };
}

export function dissociateInviteeAndClearInferredFields(): AsyncAppThunk {
  return async (dispatch, getState) => {
    const state = getState();
    const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
    // invitee associated attendeeId and contact Id
    delete eventRegistration.attendee.attendeeId;
    delete eventRegistration.attendee.personalInformation.contactId;
    // inferred information to clear: 2. inviteeId related personalInformation
    const existingContactInfo = eventRegistration.attendee.personalInformation;
    const contactInfo = await Promise.all(
      getRegistrationPageFields(state).map(async field => {
        const isFieldReviewed = await isWidgetReviewed(state, { fieldId: field.fieldId });
        return field.display !== NOT_VISIBLE && field.display !== READ_ONLY && isFieldReviewed ? field : null;
      })
    );
    const filteredContactInfo = contactInfo.filter(Boolean);
    const copiedContactInfo = filteredContactInfo.map(field =>
      field.isCustomField
        ? {
            customFields: {
              [field.fieldId]: existingContactInfo.customFields[field.fieldId]
            }
          }
        : {
            // @ts-expect-error ts-migrate(2464) FIXME: A computed property name must be of type 'string',... Remove this comment to see the full error message
            [(StandardContactFields[field.fieldId] || {}).regApiPath]:
              // @ts-expect-error ts-migrate(2538) FIXME: Type 'string[]' cannot be used as an index type.
              existingContactInfo[(StandardContactFields[field.fieldId] || {}).regApiPath]
          }
    );
    const temp = copiedContactInfo.reduce(
      (copiedInfo, info) => (info.customFields ? merge(copiedInfo, info) : Object.assign(copiedInfo, info)),
      {}
    );
    eventRegistration.attendee.personalInformation = temp;

    await dispatch(clearOutOfSyncUserSessionFieldsAndUrlQuery());
    // Reversing the order so that inviteeId is set to undefined for sourceId to be shown as editable.
    dispatch({
      type: CLEAR_REG_CART_INFERRED_FIELDS,
      payload: {
        regCart: setIn(
          getRegCart(state),
          ['eventRegistrations', eventRegistration.eventRegistrationId],
          eventRegistration
        )
      }
    });
  };
}
