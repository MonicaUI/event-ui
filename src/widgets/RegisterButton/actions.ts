import { routeToPage, redirectToPage } from '../../redux/pathInfo';
import { logoutRegistrant } from '../../redux/registrantLogin/actions';
import { openStartNewRegistrationDialogDuringRegistration } from '../../dialogs';
import * as currentRegistrant from '../../redux/selectors/currentRegistrant';
import { DECLINE, GUEST_REGISTRATION, REGISTRATION } from '../../redux/website/registrationProcesses';

/**
 * @param props.config.text Registration dialog text
 * @param props.style Styling for registration dialog
 * @param props.classes Classnames for registration dialog
 * Based on user state and current page
 * Dispatch actions either to redirect to registration start,
 * logout and redirect to 'register' page, or prompt with new registration dialog.
 * RegisterNow widget can exist in each of the conditions, all actions are available
 * RegisterAnother widget can only exists on post-reg pages,
 * openStartNewRegistrationDialog is the only available action
 */
export const promptToStartNewRegistration = (props: $TSFixMe) => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const loggedIn = currentRegistrant.isLoggedIn(state);
    const registrationPage = REGISTRATION.isTypeOfCurrentPage(state) || GUEST_REGISTRATION.isTypeOfCurrentPage(state);
    const registrationDeclinePage = DECLINE.isTypeOfCurrentPage(state);
    const titleText = props.config.text.htmlContent
      ? props.config.text.htmlContent.replace(/<[^>]*>/g, '')
      : props.config.text;
    if (!loggedIn && !registrationPage && !registrationDeclinePage) {
      return dispatch(routeToPage('register'));
    }
    if (registrationDeclinePage) {
      await dispatch(logoutRegistrant());
      return dispatch(redirectToPage('register'));
    }
    return dispatch(
      openStartNewRegistrationDialogDuringRegistration({
        title: titleText,
        classes: { ...props.classes },
        style: props.style,
        startWithDefaultRegPath: props.startWithDefaultRegPath || false
      })
    );
  };
};
