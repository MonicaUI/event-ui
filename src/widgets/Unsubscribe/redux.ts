import { injectPageSnapshot } from '../../redux/actions';
import querystring from 'querystring';

/**
 * Unsubscribe and re-subscribe Redux module:
 * Event invitees can click "Unsubscribe" link in their emails to opt them out from the event emails communication
 */
const GET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/GET_UNSUBSCRIBE_STATUS';
const SET_UNSUBSCRIBE_STATUS = 'event-guestside-site/unsubscribe/SET_UNSUBSCRIBE_STATUS';

const createGetUnsubscribeStatusAction = (unsubscribedContact, optOutStatus) => {
  return {
    type: GET_UNSUBSCRIBE_STATUS,
    payload: {
      ...optOutStatus,
      unsubscribeStatus: !unsubscribedContact.subscribed
    }
  };
};

export const getUnsubscribeStatus =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient, eventGuestClient },
      accessToken,
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const unsubscribedContact = await eventEmailClient.getSubscriptionStatus(eventId, inviteeId, encodedGuestId);
    // optOutStatus object has invitee/guest and planner infomation for display
    const optOutStatus = await eventGuestClient.getOptOutStatus(accessToken, eventId, inviteeId, encodedGuestId);
    dispatch(createGetUnsubscribeStatusAction(unsubscribedContact, optOutStatus));
  };

const createUnsubscribeAction = unsubscribeStatus => {
  return {
    type: SET_UNSUBSCRIBE_STATUS,
    payload: { unsubscribeStatus }
  };
};

export const unsubscribeAction =
  (unsubscribeStatus: $TSFixMe): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { eventEmailClient },
      event: { id: eventId },
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    const subscribed = !unsubscribeStatus;
    await eventEmailClient.setSubscriptionStatus(eventId, inviteeId, encodedGuestId, subscribed);
    dispatch(createUnsubscribeAction(unsubscribeStatus));
  };

export const prepareForUnsubscribePageLoad =
  (): $TSFixMe =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      userSession: { inviteeId }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const queryParams = window && window.location && querystring.parse(window.location.search.slice(1));
    const encodedGuestId = queryParams.g;
    let unsubscribePage;
    if (!inviteeId && !encodedGuestId) {
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './invalidInviteeContent'))
        .invalidInviteeUnsubscribePage;
    } else {
      await dispatch(getUnsubscribeStatus());
      // eslint-disable-next-line import/no-cycle
      unsubscribePage = (await import(/* webpackChunkName: "unsubscribe" */ './content')).unsubscribePage;
    }
    dispatch(injectPageSnapshot(unsubscribePage));
  };

export const unsubscribeReducer = (state = {}, action: $TSFixMe): $TSFixMe => {
  switch (action.type) {
    case GET_UNSUBSCRIBE_STATUS:
    case SET_UNSUBSCRIBE_STATUS:
      return {
        ...state,
        ...action.payload
      };
    default:
      return state;
  }
};
