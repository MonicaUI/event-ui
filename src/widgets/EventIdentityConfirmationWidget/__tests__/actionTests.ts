/* eslint-env jest */
import appData from '../../../redux/registrationForm/regCart/__tests__/appData.json';
import { clearFields, dissociateInviteeAndClearInferredFields } from '../actions';
import history from '../../../myHistory';
import { CLEAR_REG_CART_INFERRED_FIELDS } from '../../../redux/registrationForm/regCart/actionTypes';
import EventSnapshot from '../../../../fixtures/EventSnapshot.json';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { populateVisibleProducts } from '../../../redux/visibleProducts';
import createStoreWithMiddleware from 'nucleus-guestside-site/src/redux/devCreateStoreWithMiddleware';
import { combineReducers } from 'redux';
import EventSnapshotClient from '../../../clients/EventSnapshotClient';
import * as ExperimentHelper from '../../../ExperimentHelper';
import * as dialogs from '../../../dialogs';

const mockOpenPrivateEventErrorDialog = jest.spyOn(dialogs, 'openPrivateEventErrorDialog');
let mockUseGraphQLSiteEditorData = ExperimentHelper.GraphQLSiteEditorDataReleases.Development;
jest.mock('../../../ExperimentHelper', () => ({
  ...jest.requireActual<typeof ExperimentHelper>('../../../ExperimentHelper'),
  getUseGraphQLSiteEditorData: () => mockUseGraphQLSiteEditorData
}));
jest.mock('../../../apollo/siteEditor/pageVarietyPathQueryHooks', () => ({
  createPageVarietyPathManualQuery: () => ({
    data: {
      event: {
        registrationPath: {
          registration: {
            registrationType: {
              validation: {
                onPageBeforeIdentityConfirmation: false
              }
            }
          }
        }
      }
    }
  })
}));

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const eventId = EventSnapshot.eventSnapshot.id;
const regCart = {
  regCartId: '928f02df-4ba7-4637-824d-ae71bd414e4f',
  status: 'INPROGRESS',
  eventSnapshotVersions: {
    [(EventSnapshot as $TSFixMe).id]: 'fake-eventSnapshot-version'
  },
  eventRegistrations: {
    '00000000-0000-0000-0000-000000000001': {
      eventRegistrationId: '00000000-0000-0000-0000-000000000001',
      eventId,
      attendee: {
        personalInformation: {
          contactId: 'd10d9eab-743f-4b3e-80ee-adb520920281',
          firstName: 'Luke',
          lastName: 'Roling',
          emailAddress: 'lroling-384934@j.mail',
          primaryAddressType: 'WORK'
        },
        attendeeId: 'aac17ae9-74b8-4687-ae9e-e2781d247f95'
      },
      attendeeType: 'ATTENDEE',
      productRegistrations: [
        {
          productId: Object.keys(EventSnapshot.eventSnapshot.products.admissionItems)[0],
          productType: 'AdmissionItem',
          quantity: 1,
          requestedAction: 'REGISTER'
        }
      ],
      sessionRegistrations: {},
      registrationTypeId: '00000000-0000-0000-0000-000000000000',
      registrationPathId: '411c6566-1e5a-4c38-b8e5-f63ab9239b40'
    }
  }
};

jest.mock('../../../redux/visibleProducts');
const mockedVisibleProducts = populateVisibleProducts as jest.Mock;
mockedVisibleProducts.mockImplementation(() => ({
  type: '[MOCK]/LOAD_VISIBLE_SESSION_PRODUCTS',
  payload: {}
}));

jest.mock('../../../clients/EventSnapshotClient', () => {
  return {
    getVisibleProducts: jest.fn(() => ({ Sessions: {} }))
  };
});

describe.each([
  ['GraphQL', ExperimentHelper.GraphQLSiteEditorDataReleases.Development],
  ['Redux', ExperimentHelper.GraphQLSiteEditorDataReleases.Off]
])('action tests using %s site editor data', (_description, experimentStatus) => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseGraphQLSiteEditorData = experimentStatus;
  });

  test("clearInferredFields clear regCart fields that's on page or upcoming", async () => {
    const eventRegistrationId = '00000000-0000-0000-0000-000000000001';
    const initialState = {
      event: EventSnapshot.eventSnapshot,
      account: EventSnapshot.accountSnapshot,
      website: EventSnapshot.eventSnapshot.siteEditor.website,
      appData,
      registrationForm: {
        regCart
      },
      regCartStatus: {
        lastSavedRegCart: regCart
      },
      pathInfo: {
        rootPath: '/events/dummyId',
        currentPageId: 'regProcessStep1'
      },
      registrationPaths: {
        'path-0': {
          pageIds: [
            'regProcessStep1',
            'regPage:0cc99264-e900-43b2-a995-91b920ff2a33',
            'regPage:617b1dc3-3ddb-4468-8087-049a4e6e51e8'
          ],
          id: '9f632f6a-40d9-4870-beeb-8f38a8151b54',
          confirmationPageId: 'confirmation',
          postRegPageIds: ['confirmation']
        }
      },
      userSession: {
        regTypeId: 'userSessionRegTypeId',
        inviteeId: 'inviteeId'
      },
      defaultUserSession: {}
    };
    const store = mockStore(initialState);
    // @ts-expect-error ts-migrate(2554) FIXME: Expected 0 arguments, but got 1.
    // eslint-disable-next-line jest/valid-expect-in-promise
    await store.dispatch(dissociateInviteeAndClearInferredFields(eventRegistrationId));
    expect(store.getActions()).toMatchSnapshot();
    expect(history.location.pathname).toBe('/events/dummyId/regProcessStep1');
    // explicitly check personalInformation is cleared in snapshot
    const clearAction = store.getActions().find(action => action.type === CLEAR_REG_CART_INFERRED_FIELDS);
    expect(clearAction.payload.regCart.eventRegistrations[eventRegistrationId].attendee.attendeeId).toBe(undefined);
    expect(
      clearAction.payload.regCart.eventRegistrations[eventRegistrationId].attendee.personalInformation.firstName
    ).toBe(undefined);
  });

  const store = createStoreWithMiddleware(
    combineReducers({
      website: (x = {}) => x,
      text: (x = {}) => x,
      regCartStatus: (x = {}) => x,
      clients: (x = {}) => x,
      registrantLogin: (x = {}) => x,
      event: (x = {}) => x,
      userSession: (x = {}) => x,
      defaultUserSession: (x = {}) => x,
      registrationForm: (x = {}) => x,
      appData: (x = {}) => x,
      pathInfo: (x = {}) => x
    }),
    {
      website: EventSnapshot.eventSnapshot.siteEditor.website,
      appData,
      clients: {
        eventSnapshotClient: EventSnapshotClient
      },
      registrantLogin: {
        form: {
          emailAddress: 'email',
          confirmationNumber: 'confirmationNumber'
        }
      },
      regCartStatus: {
        lastSavedRegCart: regCart
      },
      event: {
        id: eventId
      },
      text: {
        translate: (resx, tokens) => (tokens ? `${resx}:${JSON.stringify(tokens)}` : resx)
      },
      userSession: {},
      defaultUserSession: {
        eventId
      },
      registrationForm: {
        regCart
      },
      pathInfo: {
        currentPageId: 'regPage1',
        rootPath: 'http://example123?'
      }
    },
    {
      thunkExtraArgument: { apolloClient: {} }
    }
  );

  const createError = responseBody => ({
    responseStatus: 422,
    responseBody,
    httpLogRequestId: 'requestHeader',
    httpLogPageLoadId: 'pageLoadId',
    errorDateTime: new Date()
  });

  test('clear fields must reevaluate visible session Products', async () => {
    await store.dispatch(clearFields());
    expect(populateVisibleProducts).toHaveBeenCalled();
    expect(populateVisibleProducts).toHaveBeenCalledWith('00000000-0000-0000-0000-000000000001');
  });

  test('clear fields handles error', async () => {
    mockedVisibleProducts.mockImplementation(() => {
      const responseBody = {
        validationMessages: [
          {
            severity: 'Error',
            localizationKey: 'REGAPI.EMAIL_ONLY_INVITEE'
          }
        ]
      };
      throw createError(responseBody);
    });

    await store.dispatch(clearFields());
    expect(mockOpenPrivateEventErrorDialog).toHaveBeenCalled();
    mockOpenPrivateEventErrorDialog.mockClear();
  });
});
