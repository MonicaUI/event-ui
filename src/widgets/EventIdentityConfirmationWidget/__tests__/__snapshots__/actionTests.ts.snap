// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action tests using GraphQL site editor data clearInferredFields clear regCart fields that's on page or upcoming 1`] = `
Array [
  Object {
    "type": "event-guestside-site/registrantLogin/CLEAR_URL_IDENTIFYING_INFORMATION",
  },
  Object {
    "payload": Object {
      "regCart": Object {
        "eventRegistrations": Object {
          "00000000-0000-0000-0000-000000000001": Object {
            "attendee": Object {
              "personalInformation": Object {},
            },
            "attendeeType": "ATTENDEE",
            "eventId": "dd48b4c4-672e-431c-be8d-5abfec299844",
            "eventRegistrationId": "00000000-0000-0000-0000-000000000001",
            "productRegistrations": Array [
              Object {
                "productId": "c0215717-5640-4e9d-b790-36047f14bf21",
                "productType": "AdmissionItem",
                "quantity": 1,
                "requestedAction": "REGISTER",
              },
            ],
            "registrationPathId": "411c6566-1e5a-4c38-b8e5-f63ab9239b40",
            "registrationTypeId": "00000000-0000-0000-0000-000000000000",
            "sessionRegistrations": Object {},
          },
        },
        "eventSnapshotVersions": Object {
          "undefined": "fake-eventSnapshot-version",
        },
        "regCartId": "928f02df-4ba7-4637-824d-ae71bd414e4f",
        "status": "INPROGRESS",
      },
    },
    "type": "event-guestside-site/regCart/CLEAR_REG_CART_INFERRED_FIELDS",
  },
]
`;

exports[`action tests using Redux site editor data clearInferredFields clear regCart fields that's on page or upcoming 1`] = `
Array [
  Object {
    "type": "event-guestside-site/registrantLogin/CLEAR_URL_IDENTIFYING_INFORMATION",
  },
  Object {
    "payload": Object {
      "regCart": Object {
        "eventRegistrations": Object {
          "00000000-0000-0000-0000-000000000001": Object {
            "attendee": Object {
              "personalInformation": Object {},
            },
            "attendeeType": "ATTENDEE",
            "eventId": "dd48b4c4-672e-431c-be8d-5abfec299844",
            "eventRegistrationId": "00000000-0000-0000-0000-000000000001",
            "productRegistrations": Array [
              Object {
                "productId": "c0215717-5640-4e9d-b790-36047f14bf21",
                "productType": "AdmissionItem",
                "quantity": 1,
                "requestedAction": "REGISTER",
              },
            ],
            "registrationPathId": "411c6566-1e5a-4c38-b8e5-f63ab9239b40",
            "registrationTypeId": "00000000-0000-0000-0000-000000000000",
            "sessionRegistrations": Object {},
          },
        },
        "eventSnapshotVersions": Object {
          "undefined": "fake-eventSnapshot-version",
        },
        "regCartId": "928f02df-4ba7-4637-824d-ae71bd414e4f",
        "status": "INPROGRESS",
      },
    },
    "type": "event-guestside-site/regCart/CLEAR_REG_CART_INFERRED_FIELDS",
  },
]
`;
